File: platform/lang-impl/testSources/com/intellij/openapi/wm/impl/HideAllToolWindowsTest.java
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.openapi.wm.impl;

import com.intellij.openapi.actionSystem.ActionManager;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.wm.ToolWindow;
import com.intellij.openapi.wm.ToolWindowEP;
import com.intellij.openapi.wm.ToolWindowType;
import com.intellij.testFramework.MapDataContext;
import gnu.trove.THashMap;
import gnu.trove.THashSet;

import java.util.*;

public class HideAllToolWindowsTest extends ToolWindowManagerTestCase {
  public void testDontHideFloatingAndWindowedToolWindows() {
    List<ToolWindowEP> extensions = ToolWindowEP.EP_NAME.getExtensionList();

    Map<String, ToolWindowType> types = new THashMap<>();
    List<ToolWindowType> cycle = new ArrayList<>(Arrays.asList(ToolWindowType.values()));
    for (int i = 0; i < extensions.size(); i++) {
      ToolWindowEP extension = extensions.get(i);
      String id = extension.id;
      if (id.equals("Structure") || id.equals("Favorites") || id.equals("Ant")) {
        continue;
      }

      manager.initToolWindow(extension);
      // if not applicable, then will be not registered
      if (!manager.isToolWindowRegistered(id)) {
        continue;
      }

      manager.showToolWindow(id);

      ToolWindowType type = cycle.get(i % cycle.size());
      types.put(id, type);
      manager.setToolWindowType(id, type);
    }

    MapDataContext context = new MapDataContext();
    context.put(CommonDataKeys.PROJECT, getProject());
    Set<String> visibleIds = new THashSet<>();
    for (String id : types.keySet()) {
      if (manager.getToolWindow(id).isVisible()) {
        visibleIds.add(id);
      }
    }
    ActionManager.getInstance().getAction("HideAllWindows").actionPerformed(AnActionEvent.createFromDataContext("", null, context));

    for (String id : visibleIds) {
      ToolWindow window = manager.getToolWindow(id);
      assertEquals(id + ":" + window.getType(),
                   types.get(id) == ToolWindowType.FLOATING || types.get(id) == ToolWindowType.WINDOWED,
                   window.isVisible());
    }
  }
}

File: platform/lang-impl/testSources/com/intellij/toolWindow/HideAllToolWindowsTest.kt
// Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.toolWindow

import com.intellij.openapi.actionSystem.ActionManager
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.application.EDT
import com.intellij.openapi.wm.ToolWindowEP
import com.intellij.openapi.wm.ToolWindowId
import com.intellij.openapi.wm.ToolWindowType
import com.intellij.testFramework.MapDataContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext

class HideAllToolWindowsTest : ToolWindowManagerTestCase() {
  fun testDontHideFloatingAndWindowedToolWindows() {
    runBlocking {
      withContext(Dispatchers.EDT) {
        val extensions = ToolWindowEP.EP_NAME.extensionList
        val types: MutableMap<String, ToolWindowType> = HashMap()
        val cycle = listOf(*ToolWindowType.values())
        for (i in extensions.indices) {
          val extension = extensions[i]
          val id = extension.id
          if (id == "Structure" || id == ToolWindowId.FAVORITES_VIEW || id == ToolWindowId.BOOKMARKS || id == "Ant") {
            continue
          }
          manager!!.initToolWindow(extension)
          // if not applicable, then will be not registered
          if (!manager!!.isToolWindowRegistered(id)) {
            continue
          }
          manager!!.showToolWindow(id)
          val type = cycle[i % cycle.size]
          types[id] = type
          manager!!.setToolWindowType(id, type)
        }
        val context = MapDataContext()
        context.put(CommonDataKeys.PROJECT, project)
        val visibleIds: MutableSet<String> = HashSet()
        for (id in types.keys) {
          if (manager!!.getToolWindow(id)!!.isVisible) {
            visibleIds.add(id)
          }
        }
        ActionManager.getInstance().getAction("HideAllWindows").actionPerformed(AnActionEvent.createFromDataContext("", null, context))
        for (id in visibleIds) {
          val window = manager!!.getToolWindow(id)
          assertEquals(id + ":" + window!!.type,
                       types[id] == ToolWindowType.FLOATING || types[id] == ToolWindowType.WINDOWED,
                       window.isVisible)
        }
      }
    }
  }
}